include/tvm/compilation.h

```diff
@@ -0,0 +1,211 @@
+/*!
+*  Copyright (c) 2017 by Contributors
+* \file compilation.h
+* \brief Functions for compiling ops.
+*/
+#ifndef TVM_COMPILATION_H_
+#define TVM_COMPILATION_H_
+
+#include <string>
+#include "./tvm/c_dsl_api.h"
+#include "./tvm/runtime/packed_func.h"
+#include "./tvm/schedule_pass.h"
+#include "./tvm/lowered_func.h"
+
+namespace tvm {
+
+namespace compilation {
+
+/*!
+* \brief Container for target device information.
+* Use target_llvm, target_cuda etc functions instead of constructing directly.
+*/
+struct Target {
+    /*! \brief The name of the target device */
+    std::string targetName;
+    /*! \brief The type of the target device */
+    DLDeviceType deviceType;
+    /*! \brief The maximum threads that a schedule should use for this device */
+    int max_num_threads;
+    /*! \brief The warp size that should be used by the LowerThreadAllreduce pass */
+    int thread_warp_size;
+    /*! \brief Keys for this target */
+    std::unordered_set<std::string> keys;
+    /*! \brief Options for this target */
+    std::unordered_set<std::string> options;
+
+
+    Target(std::string targetName, DLDeviceType deviceType, int max_num_threads,
+        int thread_warp_size, const std::unordered_set<std::string>& keys,
+        const std::unordered_set<std::string>& options) {
+        this->targetName = targetName;
+        this->deviceType = deviceType;
+        this->max_num_threads = max_num_threads;
+        this->thread_warp_size = thread_warp_size;
+        this->keys = keys;
+        this->options = options;
+    }
+
+    /*! \return the full device string to pass to codegen::Build */
+    std::string str() const {
+        std::ostringstream result;
+        result << targetName;
+        for (const auto &x : options) {
+            result << " " << x;
+        }
+        return result.str();
+    }
+
+    /*! \return True iff this target has the given key */
+    bool hasKey(std::string key) const {
```

Comment:
Since this is short function, do not use it. Always pass string by const reference when possible.